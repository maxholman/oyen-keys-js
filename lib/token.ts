import type { JsonObject, SimplifyDeep } from 'type-fest';
import { base64UrlToObject, uint8ArrayToBase64Url } from './base64.js';
import { TokenExpiredError } from './errors.js';
import {
  getImportParams,
  getSigningParamsFromJwk,
  type JwtAlg,
} from './jwk.js';
import { decode, encode, type JwtPayload } from './jwt.js';
import {
  assertKeyInObjectIsString,
  assertString,
  assertTruthy,
  inline,
} from './utils.js';

async function definitelyJwk<T extends JsonWebKey>(
  maybeJwk: string | T,
  keyUsages: ReadonlyArray<KeyUsage>,
) {
  const jwk =
    typeof maybeJwk === 'string'
      ? base64UrlToObject<SimplifyDeep<T>>(maybeJwk)
      : maybeJwk;

  const importParams = getImportParams(jwk);

  return {
    key: await crypto.subtle.importKey(
      'jwk',
      jwk,
      importParams,
      false,
      keyUsages,
    ),
    jwk,
  };
}

export async function signToken<T extends JwtPayload>(
  maybeJwk: string | JsonWebKey,
  {
    kid,
    claims,
    ttlSecs = 60 * 60,
  }: {
    kid: string;
    claims: Omit<T, 'exp' | 'iat'> & Partial<Pick<T, 'exp' | 'iat'>>;
    ttlSecs: number;
  },
) {
  assertString(kid);

  const { jwk, key } = await definitelyJwk(maybeJwk, ['sign']);

  const alg = inline((): JwtAlg => {
    switch (jwk.kty) {
      case 'EC':
        return 'ES256';
      case 'RSA':
        return 'RS256';
      case 'OKP':
        return 'EdDSA';
      default:
        throw new Error(`Unsupported key type ${JSON.stringify(jwk.kty)}`);
    }
  });

  const header = {
    typ: 'JWT',
    kid,
    alg,
  };

  const iat =
    claims && 'iat' in claims && typeof claims.iat === 'number'
      ? claims.iat
      : Math.floor(Date.now() / 1000);

  const payload = {
    ...claims,
    iat,
    exp: iat + ttlSecs,
  };

  const encodedData = encode(header, payload);
  const signParams = getSigningParamsFromJwk(alg, jwk);

  const signature = await crypto.subtle.sign(
    signParams,
    key,
    new TextEncoder().encode(encodedData),
  );

  return `${encodedData}.${uint8ArrayToBase64Url(new Uint8Array(signature))}`;
}

export function decodeToken<T extends JwtPayload | JsonObject>(jwt: string) {
  return decode<T>(jwt);
}

export async function verifyToken<T extends JwtPayload | JsonObject>(
  maybeJwk: string | JsonWebKey,
  jwt: string,
) {
  const { header, payload, signature, signedData } = decodeToken<T>(jwt);

  if (
    'nbf' in payload &&
    typeof payload.nbf === 'number' &&
    payload.nbf > Math.floor(Date.now() / 1000)
  ) {
    throw new TokenExpiredError('Not yet valid').debug({
      nbf: payload.nbf,
      payload,
    });
  }

  if (
    'exp' in payload &&
    typeof payload.exp === 'number' &&
    payload.exp <= Math.floor(Date.now() / 1000)
  ) {
    throw new TokenExpiredError('Expired').debug({
      exp: payload.exp,
      payload,
    });
  }

  // must have an alg we know of
  assertKeyInObjectIsString(header, 'alg');

  // paranoia
  assertTruthy(header.alg !== 'none');

  const { jwk, key } = await definitelyJwk(maybeJwk, ['verify']);

  const verified = await crypto.subtle.verify(
    getSigningParamsFromJwk(header.alg, jwk),
    key,
    signature,
    signedData,
  );

  if (!verified) {
    throw new Error('Token did not verify');
  }
  return { header, payload };
}
