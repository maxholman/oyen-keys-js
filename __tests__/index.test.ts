import { describe, expect, test, vi } from 'vitest';
import { createToken } from '../lib/token.js';

// fake for tests
const privateKey =
  'eyJhbGciOiJSUzI1NiIsImt0eSI6IlJTQSIsImUiOiJBUUFCIiwiZCI6IlEzSFh4bjBmWUJIV2U2NUdQVGxiV0k2OFBfQTZhb0YzYWV2dGdmX3JJQVJfeThjWE9tNkRXOTJBbGhTc2pwcU1HeUo1US1zMDVicV9oS2lVRGg0Sk1URFROdkZKWkZOY3VnUzZ1bXRGLWpiRVp1c1RxMzlGU3NyUTJIcEVFVy1sa2ZlVVZ6UUJ4ZXdRUE1oemVuVWY3ZFZ6dkpSMUhqVkVZRFJBckpFOEUtd2x5UXNlNUgyOENzM0dOUEt1OXBtZXlrTFBQazhrYU5SZ0hoMWhCc21kTk5zSzhCa2VSWFJXZEVDQnpPc0tsakVFUWdiZUxWam5SNGt3b3FpbGJybExOX3BVY3Q5RHZ4LTdkNEtGNDRZODNwdWZmOFVTMmgtemp4eW11ZFZyd3pxNHliYkR2SHlWY0hORjV4WnlQNDcxNWpoRnNaaXAyVTkzdS1RMVFRSXRKUSIsIm4iOiJ5Uzg0NVFxOFdGQWFLaTFvS0RmTDNqWE9mNkU2RTRRRm0tX01EZm5IZjY4V0gtRG1va1l6WnIxaWtLeVN5SWdzWS04bjZDUnBDcnV4YjNWTmxNcUczTjBXN0M0VGExcE1OVndacGRORnk1WWRLVWhpRjliYXVGN2p1dGpFcnAwRnN2dHJKTEM1dFdQeVlJVjgtLVpTMG1TS1BWT2c3WExkeENDMHFYY25xTTVkemZZczRnX3hrRFN5VFhSeFB4WVhzR0Z1aUV5VC1kM1NUXzRaWVFzN2x2bHAwdE5QQUJMUFNoRS14OE81aTVSemFhU0hjM0o5UmdseDMyemVuMEpnQkJMSTRiMmZEbmdEX0FJRnRJS1BabmNseUE0LTlBS1dGdGJPTi1ySndrZXFUOGxEOGN3NElFQkxVUVNGR2dvZmZKbXVtempEdGpKcVZCMThjRW5PT3ciLCJwIjoiX2ZkSzBpUHZSUGFsTXFMbmpWVTRBZjBycU9nNHRHeXIxbjQ4MXl4S3QyS1daeXV6NFp2Z1dXUDI4R05mVkJGWC1FcnQyaF9jZnIyNXRrNVFBSVkyVXVvZlFxQ3NSRzdlTm5xOXNsdXJISmlUQkhDTFluUGE0UlJaY2xna1A1YmVMbU1VLTRJTDBkSVJSVHcyR2ZzTDBsUUs5WlFkSlNXRE1yRG00UU9WRnhjIiwicSI6InlzdTJOWU5tYnkwS3VXTTMyV3JwalRYRS1CSTZTQTc5UDFUQ0hRZEl5VmVSS1JSV0VLUUdncU9sWnJYeVAyVmxQWHg4d0syT1NQWFNzWFhqODg2WUZHRjFmZGJPRkhIZmNKYlVPcE1KTmFJVWFqSjMtaG1sYUJYaHlJN25PUEViRGVuQzhSV2dJdG9GeFlmb0N3YnQtTHpZTTRKWXlBUjE4b0lJOG5aanVIMCIsImRwIjoiNmhrYkZOMGUzdGpqcWV0VHVZbUdHUDhfVVhETlpES1BxeVd0LTBIT2xZQUExNVVIT0wwd2JmV2hXRENoWEtLb0RpMHpWazdzTnItNEFmM3A3TEhLUFYtbDlRcWFBSEtaem1vUmxlaTlyNXhNV0ZsTmdqRDgzMkNjbXJQNHBjenhmSnpFYUptVlA1RjliZTUwaURwS3lWUEN6bF83QlpTd2ZQZG51cDVNdlgwIiwiZHEiOiJaSFk2VTVRLW9JQzl1VTB2azZObEE3RkpYR01iUUNzOHhhemxaS0FCTE0wV2VqVTYxWS1NTWhlRGpyZ193NndsVlNwZ1dzc2ZtNEU3MlBHXzlUMjVGUS1hdHJZODdOX2tndEFWWHRjTVR1ZDJHWTNhNWFqaXlnZGRESV9PZC1UcE5DRkRrMDdWUDY2bUFVN3NZdG9SOVpFSHdheW04cDlURnNuWVNOVGhYYmsiLCJxaSI6IjhiNTRpdGUxWGtnNkxJa2FxdDgzTE80RG13eTVld3NYWjlBOGJXdmZQSEFuZEY1LWgzTkQyMzdOOUpnaGY1d2hzR09tRVZhTWd2NTdSZU1SakJZRGs0RWVMeWhvOW5jemY5ZU5fbmdFd3JkbWxjQ1NtV1NPNG9fdHdDaGc2b0hwN05WRkJBLWJsSjJmb1ZfNkxYNnl6RE9mcFVfeWR6ZmJkeWV1eXI5WGZmbyIsImtpZCI6ImNYc2hWTndTdkRaOCJ9';

describe('Basic', () => {
  vi.setSystemTime(new Date('2021-01-01T00:00:00Z'));

  test('Nothing', async () => {
    const staticToken = await createToken({
      privateKey,
      claims: {
        sub: 'me',
        cap: {
          'team:123': ['read', 'publish'] as const,
        },
      },
      ttlSecs: 30,
    });

    expect(staticToken).toMatchSnapshot();
  });
});
