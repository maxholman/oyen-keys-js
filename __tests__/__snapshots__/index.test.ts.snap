// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Basic > Nothing 1`] = `"eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6ImNYc2hWTndTdkRaOCJ9.eyJzdWIiOiJtZSIsImNhcCI6eyJ0ZWFtOjEyMyI6WyJyZWFkIiwicHVibGlzaCJdfSwiaWF0IjoxNjA5NDU5MjAwLCJleHAiOjE2MDk0NTkyMzB9.Loky1NktDe4flcm8xEQmE0nSM1qjnGprvRmT_6VjS4YU97K2OK8rHEwV_TbUpH9pcFT_DJoggQbPvZgk6NFWuKQQaMY74tGap8F4Lx7ZjCFzqbQoFUSVEguG57S5_JFfv-Sypfwi6MsA39V2l801fF_lhPjZhmXH_c_sw2m87bVzlZCEyOCcuTg-6BIex_0eLMZp8lEwssVZIvs199GoowKsKAvZF8rPZ82W92UaRvBPwIzdA8H2yVzJmpi358VIagfF19NroZiO0-Ubc7iWAl_hIWaxTrrUByxuSvqjrHfsuzIU3NRLEUvxLsuIIgMSvCKjZwfoE-bZi4OffnfXiA"`;
